File: CivitasRuby/lib/mazo_sorpresas.rb
require_relative 'sorpresa'

module Civitas

  class MazoSorpresas
    
    attr_reader :ultima_sorpresa
    
    def init
      @sorpresas=Array.new
      @cartas_especiales=[]
      @barajada=false
      @usadas=0
      @ultima_sorpresa=nil
    end
    
    def initialize(debug=false)
      @debug=debug
      init
      if(@debug)
        Diario.instance.ocurre_evento("El modo debug ha sido activado") 
      end
    end
    
    def al_mazo(s)
      if(!@barajada)
        @sorpresas << s
      end
    end
    
    def siguiente
      aux = nil
      if(@barajada==false)
          @usadas=0
          @sorpresas.shuffle!
          @barajada=true
      end
      @usadas+=1
      aux=@sorpresas[0]
      @sorpresas.shift
      @sorpresas << aux
      @ultima_sorpresa=aux

      return @ultima_sorpresa
    end
    
    def inhabilitar_carta_especial(sorpresa)
      encontrado=false
      posicion_encontrado=0
      i=0
      while (i<@sorpresas.length || encontrado==false)
        if(@sorpresas[i] == sorpresa)
          encontrado=true
          posicion_encontrado=i
        end 
        i += 1
      end
      
      if(encontrado==true)
        @sorpresas.delete_at(posicion_encontrado)
        @cartas_especiales << sorpresa
        Diario.instance.ocurre_evento("Se ha inhabilitado una carta especial")
      end
    end

    def habilitar_carta_especial(sorpresa)
      encontrado=false
      posicion_encontrado=0
      i=0
      while (i<@cartas_especiales.length || encontrado==false)
        if(@cartas_especiales[i] == sorpresa)
          encontrado=true
          posicion_encontrado=i
        end  
        i += 1
      end

      if(encontrado==true)
        @cartas_especiales.delete_at(posicion_encontrado)
        @sorpresas << sorpresa
        Diario.instance.ocurreEvento("Se ha habilitado una carta especial")
      end
    end

    private :init
    
  end
end

File: CivitasJava/src/civitas/MazoSorpresas.java
package civitas;

import java.util.*;

public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        this.sorpresas = new ArrayList<Sorpresa>();
        this.cartasEspeciales = new ArrayList<Sorpresa>();
        this.barajada=false;
        this.usadas=0;
    }
    
    MazoSorpresas(boolean debug){
        this.debug=debug;
        init();
        if(debug){
            Diario.getInstance().ocurreEvento("El modo debug esta activado\n");
        }
    }
    
    MazoSorpresas(){
        debug=false;
        init();
    }
    
    void alMazo(Sorpresa s){
        if(!barajada){
            sorpresas.add(s);
        }
    }
    
    Sorpresa siguiente(){
        if(!barajada){
            usadas=0;
            Collections.shuffle(sorpresas);
            barajada = true;
        }
        usadas++;
        sorpresas.add(sorpresas.get(0));
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        boolean encontrado=false;
        int posicionEncontrado=0;
        for(int i=0;i<sorpresas.size() || !encontrado;i++){
            if(sorpresas.get(i) == sorpresa){
                encontrado=true;
                posicionEncontrado=i;
            }
        }
        if(encontrado){
            sorpresas.remove(posicionEncontrado);
            cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha inhabilitado una carta especial\n");
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        boolean encontrado=false;
        int posicionEncontrado=0;
        for(int i=0;i<cartasEspeciales.size() || !encontrado;i++){
            if(cartasEspeciales.get(i) == sorpresa){
                encontrado=true;
                posicionEncontrado=i;
            }
        }
        if(encontrado){
            cartasEspeciales.remove(posicionEncontrado);
            sorpresas.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha habilitado una carta especial\n");
        }
    }
    
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
}
